README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
[![Travis build status](https://travis-ci.org/venelin/BinaryPoissonUsingSPLITT.svg?branch=master)](https://travis-ci.org/venelin/BinaryPoissonUsingSPLITT) [![Coverage status](https://codecov.io/gh/venelin/BinaryPoissonUsingSPLITT/branch/master/graph/badge.svg)](https://codecov.io/github/venelin/BinaryPoissonUsingSPLITT?branch=master) [![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/BinaryPoissonUsingSPLITT?color=blue)](http://cran.r-project.org/web/packages/BinaryPoissonUsingSPLITT) [![Downloads](http://cranlogs.r-pkg.org/badges/BinaryPoissonUsingSPLITT?color=blue)](http://cran.rstudio.com/package=BinaryPoissonUsingSPLITT)

BinaryPoissonUsingSPLITT
========================

Likelihood calculation of the binary Poisson model using the SPLITT library. This package implementats the example of a discrete model from the article 'Parallel Likelihood Calculation for Phylogenetic Comparative Models: the 'SPLITT' C++ Library'. A preprint of the article is available from <https://doi.org/10.1101/235739>. The function 'MiniBenchmark' allows to compare the calculation times for different tree sizes. See [this guide](https://venelin.github.io/SPLITT/articles/SPLITTRcppModules.html) for a tutorial.

Installation
------------

You can install the released version of BinaryPoissonUsingSPLITT from [CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("BinaryPoissonUsingSPLITT")
```

Examples
--------

-   Calculating the likelihood of the PMM for a given tree, data and model parameters.

``` r
library(ape)
library(BinaryPoissonUsingSPLITT)
#> Loading required package: Rcpp

N <- 100

x <- 1
q01 <- 0.02
q10 <- 0.8

tree <- rtree(N)
x <- sample(c(0, 1), size = N, replace = TRUE)

cat("logLikelihood using R:", BinaryPoissonModelLogLik(x, tree, 1, 0.02, 0.8), "\n")
#> logLikelihood using R: 1.537735e-42
cat("logLikelihood using R:", BinaryPoissonModelLogLikCpp(x, tree, 1, 0.02, 0.8), "\n")
#> logLikelihood using R: 1.537735e-42
```

-   Performing a benchmark to measure the likelihood calculation times using different parallelization strategies:

``` r
# N specifies the size of the phylogenetic tree. 
# Ntests specifies the number of executions in one time measurement 
# (the more the better, but also slower).
MiniBenchmark(N = 100, Ntests = 100)
#> Performing a mini-benchmark of the PMM log-likelihood calculation with 
#>       a tree of size N= 100 ;
#> Calling each likelihood calculation Ntests= 100  times ...
#> CPU:  Intel(R) Core(TM) i7-4850HQ CPU @ 2.30GHz 
#> OpenMP version:  201107 
#> Number of threads: 8 
#> Measuring calculation times...
#>    model                                            mode time.ms
#> 1    PMM                                      R (serial)    2.90
#> 2    PMM                                      C++ (AUTO)    0.04
#> 3    PMM              C++ (SINGLE_THREAD_LOOP_POSTORDER)    0.03
#> 4    PMM                 C++ (SINGLE_THREAD_LOOP_PRUNES)    0.04
#> 5    PMM                 C++ (SINGLE_THREAD_LOOP_VISITS)    0.05
#> 6    PMM                  C++ (MULTI_THREAD_LOOP_PRUNES)    0.08
#> 7    PMM                  C++ (MULTI_THREAD_LOOP_VISITS)    0.09
#> 8    PMM C++ (MULTI_THREAD_LOOP_VISITS_THEN_LOOP_PRUNES)    0.13
#> 9    PMM                  C++ (MULTI_THREAD_VISIT_QUEUE)    1.67
#> 10   PMM     C++ (MULTI_THREAD_LOOP_PRUNES_NO_EXCEPTION)    0.17
#> 11   PMM                        C++ (HYBRID_LOOP_PRUNES)    0.33
#> 12   PMM                        C++ (HYBRID_LOOP_VISITS)    0.36
#> 13   PMM       C++ (HYBRID_LOOP_VISITS_THEN_LOOP_PRUNES)    0.27
MiniBenchmark(N = 1000, Ntests = 10)
#> Performing a mini-benchmark of the PMM log-likelihood calculation with 
#>       a tree of size N= 1000 ;
#> Calling each likelihood calculation Ntests= 10  times ...
#> CPU:  Intel(R) Core(TM) i7-4850HQ CPU @ 2.30GHz 
#> OpenMP version:  201107 
#> Number of threads: 8 
#> Measuring calculation times...
#>    model                                            mode time.ms
#> 1    PMM                                      R (serial)     3.0
#> 2    PMM                                      C++ (AUTO)     0.1
#> 3    PMM              C++ (SINGLE_THREAD_LOOP_POSTORDER)     0.1
#> 4    PMM                 C++ (SINGLE_THREAD_LOOP_PRUNES)     0.0
#> 5    PMM                 C++ (SINGLE_THREAD_LOOP_VISITS)     0.1
#> 6    PMM                  C++ (MULTI_THREAD_LOOP_PRUNES)     0.5
#> 7    PMM                  C++ (MULTI_THREAD_LOOP_VISITS)     0.2
#> 8    PMM C++ (MULTI_THREAD_LOOP_VISITS_THEN_LOOP_PRUNES)     0.6
#> 9    PMM                  C++ (MULTI_THREAD_VISIT_QUEUE)     1.8
#> 10   PMM     C++ (MULTI_THREAD_LOOP_PRUNES_NO_EXCEPTION)     0.3
#> 11   PMM                        C++ (HYBRID_LOOP_PRUNES)     0.3
#> 12   PMM                        C++ (HYBRID_LOOP_VISITS)     0.3
#> 13   PMM       C++ (HYBRID_LOOP_VISITS_THEN_LOOP_PRUNES)     0.3
MiniBenchmark(N = 10000, Ntests = 10)
#> Performing a mini-benchmark of the PMM log-likelihood calculation with 
#>       a tree of size N= 10000 ;
#> Calling each likelihood calculation Ntests= 10  times ...
#> CPU:  Intel(R) Core(TM) i7-4850HQ CPU @ 2.30GHz 
#> OpenMP version:  201107 
#> Number of threads: 8 
#> Measuring calculation times...
#>    model                                            mode time.ms
#> 1    PMM                                      R (serial)     3.0
#> 2    PMM                                      C++ (AUTO)     0.1
#> 3    PMM              C++ (SINGLE_THREAD_LOOP_POSTORDER)     0.0
#> 4    PMM                 C++ (SINGLE_THREAD_LOOP_PRUNES)     0.0
#> 5    PMM                 C++ (SINGLE_THREAD_LOOP_VISITS)     0.1
#> 6    PMM                  C++ (MULTI_THREAD_LOOP_PRUNES)     0.4
#> 7    PMM                  C++ (MULTI_THREAD_LOOP_VISITS)     0.2
#> 8    PMM C++ (MULTI_THREAD_LOOP_VISITS_THEN_LOOP_PRUNES)     0.1
#> 9    PMM                  C++ (MULTI_THREAD_VISIT_QUEUE)     1.7
#> 10   PMM     C++ (MULTI_THREAD_LOOP_PRUNES_NO_EXCEPTION)     0.0
#> 11   PMM                        C++ (HYBRID_LOOP_PRUNES)     0.1
#> 12   PMM                        C++ (HYBRID_LOOP_VISITS)     0.4
#> 13   PMM       C++ (HYBRID_LOOP_VISITS_THEN_LOOP_PRUNES)     0.2
```


NEWS.md
# BinaryPoissonUsingSPLITT 1.0.1

* First release on CRAN.


tests/test-all.R
library(testthat)

test_check("BinaryPoissonUsingSPLITT")


tests/testthat/test-BinaryPoissonModelLogLikelihoodSpeed.R
library(testthat)
context("Test MiniBenchmark(1000, 10) passes without errors")


library(BinaryPoissonUsingSPLITT)

test_that("MiniBenchmark runs without errors", {
          expect_output(df <- MiniBenchmark(1000, 10), "Measuring calculation times...")
          expect_s3_class(df, "data.frame")
          })


tests/testthat/test-BinaryPoissonModelLogLikelihoodValue.R
library(testthat)
context("Test R and Cpp code calculate the same BinaryPoissonModel log-likelihood")

library(ape)
library(BinaryPoissonUsingSPLITT)


N <- 100

tree <- rtree(N)
x <- sample(c(0, 1), size = N, replace = TRUE)


test_that(
  "BinaryPoissonModelLogLik == BinaryPoissonModelLogLikCpp",
  expect_equal(BinaryPoissonModelLogLik(x, tree, 1, 0.02, 0.8),
               BinaryPoissonModelLogLikCpp(x, tree, 1, 0.02, 0.8))
)


R/zzz.R
# zzz.R
# BinaryPoissonUsingSPLITT
# 
# Copyright 2018 Venelin Mitov
# 
# This file is part of SPLITT: a generic C++ library for Serial and Parallel
# Lineage Traversal of Trees.
# 
# SPLITT is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as
# published by the Free Software Foundation, either version 3 of
# the License, or (at your option) any later version.
# 
# SPLITT is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public
# License along with SPLITT.  If not, see
# <http://www.gnu.org/licenses/>.
# 
# @author Venelin Mitov


#' Rcpp module for the \code{TraversalTaskBinaryPoissonModel}-class
#' @name BinaryPoissonUsingSPLITT__TraversalTaskBinaryPoissonModel
#' @aliases Rcpp_BinaryPoissonUsingSPLITT__TraversalTaskBinaryPoissonModel-class
NULL

#' \code{TraversalAlgorithm}-type used in \code{BinaryPoissonModel}
#' @name BinaryPoissonUsingSPLITT__BinaryPoissonModel__AlgorithmType
#' @aliases Rcpp_BinaryPoissonUsingSPLITT__BinaryPoissonModel__AlgorithmType-class
NULL

#' Base class for \code{BinaryPoissonUsingSPLITT::BinaryPoissonModel::AlgorithmType}
#' @name BinaryPoissonUsingSPLITT__BinaryPoissonModel__TraversalAlgorithm
#' @aliases Rcpp_BinaryPoissonUsingSPLITT__BinaryPoissonModel__TraversalAlgorithm-class 
#' @slot algorithm algorithm.
NULL

# loading the RCPP C++ modules
loadModule( "BinaryPoissonUsingSPLITT__TraversalTaskBinaryPoissonModel", TRUE )


R/BinaryPoissonModel.R
# BinaryPoissonModel.R
# BinaryPoissonUsingSPLITT
# 
# Copyright 2018 Venelin Mitov
# 
# This file is part of SPLITT: a generic C++ library for Serial and Parallel
# Lineage Traversal of Trees.
# 
# SPLITT is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as
# published by the Free Software Foundation, either version 3 of
# the License, or (at your option) any later version.
# 
# SPLITT is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public
# License along with SPLITT.  If not, see
# <http://www.gnu.org/licenses/>.
# 
# @author Venelin Mitov


#' Calculate the likelihood of a binary trait Poisson substituion model, 
#' given a tree and binary trait data at the tips. 
#' 
#' @description The binary trait Poisson substitution model assumes that a 
#' trait (character) with two possible states (0 and 1) changes it's state at 
#' random with constant rates of substitution from 0 to 1 (q01) and from
#' 1 to 0 (q10). The model parameters are the root-state x0, and the two rate
#' parameters q01 and q10. The likelihood for given model parameters <x0, q01, q10>
#' is defined as the probability of observing a given vector of states at the 
#' tips of a fixed phylogenetic tree. 
#' 
#' @param x an integer vector of size N, where N is the number of tips in tree.
#' @param tree a phylo object
#' @param x0,q01,q10 model parameters.
#' @param ord indices of the rows in tree$edge in pruning order. Defaults to 
#' \code{reorder(tree, order = "postorder", index.only = TRUE))}
#' 
#' @return the likelihood value.
BinaryPoissonModelLogLik <- function(
  x, tree, x0, q01, q10, 
  ord = reorder(tree, order = "postorder", index.only = TRUE)) {
  
  # number of tips in the tree
  N <- length(tree$tip.label)
  # total number of nodes in the tree (tips, internal nodes and root node)
  M <- nrow(tree$edge) + 1L
  # state variables for each node
  L0 <- L1 <- rep(1.0, M)
  
  # rate matrix
  Q <- rbind(c(-q01, q01),
             c(q10, -q10))
  
  eigQ <- eigen(Q)
  lambdaQ <- eigQ$values
  vecQ <- eigQ$vectors
  vecQ_1 <- solve(vecQ)
  
  for(o in ord) {
    # daughter node
    i <- tree$edge[o, 2]
    # parent node
    j <- tree$edge[o, 1]
    # branch length
    t <- tree$edge.length[o]
    
    Pt <- vecQ %*% diag(exp(lambdaQ * t)) %*% vecQ_1
    
    if(i <= N) {
      # initialize a tip node
      L0[i] <- as.double(x[i] == 0)
      L1[i] <- as.double(x[i] == 1)
    }
    
    L0[j] <- L0[j] * (Pt[1, 1]*L0[i] + Pt[1, 2]*L1[i])
    L1[j] <- L1[j] * (Pt[2, 1]*L0[i] + Pt[2, 2]*L1[i])
  }
  
  # for phylo objects, N+1 denotes the root node
  if(x0 == 0) {
    L0[N+1]
  } else {
    L1[N+1]
  }
}

#' Calculate the likelihood of a Poisson binary trait substituion model, 
#' given a tree and binary trait data at the tips.
#'  
#' @inheritParams BinaryPoissonModelLogLik
#' @param cppObject a previously created object returned by \code{\link{NewBinaryPoissonModelCppObject}}
#' @param mode an integer denoting the mode for traversing the tree, i.e. serial vs parallel.
#' 
#' @return the likelihood value.
BinaryPoissonModelLogLikCpp <- function(x, tree, x0, q01, q10, 
                         cppObject = NewBinaryPoissonModelCppObject(x, tree),
                         mode = getOption("SPLITT.postorder.mode", 0)) {
  L0L1 <- cppObject$TraverseTree(c(q01, q10), mode)
  
  if(x0 == 0) {
    L0L1[1]
  } else {
    L0L1[2]
  }
}


#' Create an instance of the Rcpp module for a given tree and trait data
#'
#' @inheritParams BinaryPoissonModelLogLik
#' @return an object to be passed as argument of the \link{BinaryPoissonModelLogLikCpp} function.
#' @seealso \link{BinaryPoissonModelLogLikCpp}
NewBinaryPoissonModelCppObject <- function(x, tree) {
  BinaryPoissonUsingSPLITT__TraversalTaskBinaryPoissonModel$new(tree, x[1:length(tree$tip.label)])
}


src/NumericTraitData.h
/**
 *  BinaryPoissonModel.h
 *  SPLITT
 *
 * Copyright 2017 Venelin Mitov
 *
 * This file is part of SPLITT: a generic C++ library for Serial and Parallel
 * Lineage Traversal of Trees.
 *
 * SPLITT is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * SPLITT is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with SPLITT.  If not, see
 * <http://www.gnu.org/licenses/>.
 *
 * @author Venelin Mitov
 */
#ifndef NumericTraitData_H_
#define NumericTraitData_H_

#include "./SPLITT.h"
#include <iostream>

namespace BinaryPoissonUsingSPLITT {

using namespace SPLITT;

template<class NameType>
struct NumericTraitData {
  // use const references to avoid copying of long vectors
  std::vector<NameType> const& names_;
  vec const& x_;
  NumericTraitData(
    std::vector<NameType> const& names,
    vec const& x): names_(names), x_(x) {}
};
}
#endif //NumericTraitData_H_


src/DiscreteTraitData.h
/**
 *  DiscreteTraitData.h
 *  BinaryPoissonUsingSPLITT
 *
 * Copyright 2018 Venelin Mitov
 *
 * This file is part of BinaryPoissonUsingSPLITT.
 *
 * BinaryPoissonUsingSPLITT is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * BinaryPoissonUsingSPLITT is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with SPLITT.  If not, see
 * <http://www.gnu.org/licenses/>.
 *
 * @author Venelin Mitov
 */
#ifndef DiscreteTraitData_H_
#define DiscreteTraitData_H_

#include "./SPLITT.h"

namespace BinaryPoissonUsingSPLITT {

using namespace SPLITT;

template<class NameType>
struct DiscreteTraitData {
  // use const references to avoid copying of long vectors
  std::vector<NameType> const& names_;
  uvec const& x_;
  DiscreteTraitData(
    std::vector<NameType> const& names,
    uvec const& x): names_(names), x_(x) {}
};
}
#endif //DiscreteTraitData_H_


src/BinaryPoissonModel.h
/**
 *  BinaryPoissonModel.h
 *  SPLITT
 *
 * Copyright 2018 Venelin Mitov
 *
 * This file is part of SPLITT: a generic C++ library for Serial and Parallel
 * Lineage Traversal of Trees.
 *
 * SPLITT is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * SPLITT is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with SPLITT.  If not, see
 * <http://www.gnu.org/licenses/>.
 *
 * @author Venelin Mitov
 */
#ifndef BinaryPoissonModel_H_
#define BinaryPoissonModel_H_

#include "./SPLITT.h"
#include "./DiscreteTraitData.h"
#include <iostream>
#include <armadillo>
#include <cmath>

namespace BinaryPoissonUsingSPLITT {

using namespace SPLITT;

template<class Tree>
class BinaryPoissonModel: public TraversalSpecification<Tree> {

public:
  typedef BinaryPoissonModel<Tree> MyType;
  typedef TraversalSpecification<Tree> BaseType;
  typedef Tree TreeType;
  typedef PostOrderTraversal<MyType> AlgorithmType;
  typedef vec ParameterType;
  typedef DiscreteTraitData<typename TreeType::NodeType> DataType;
  typedef vec StateType;

  // instantaneous rate parameters of transition from 0 to 1 and vice-versa
  double q01, q10;
  
  // rate matrix
  arma::mat22 Q; 
  arma::vec2 lambdaQ;
  arma::mat22 vecQ;
  arma::mat22 vecQ_1;
  
  // binary-trait vector at the tips of the tree.
  uvec x;
  
  // node-states 
  vec L0, L1;
  
  BinaryPoissonModel(TreeType const& tree, DataType const& input_data):
    BaseType(tree) {

    if(input_data.x_.size() != this->ref_tree_.num_tips()) {
      std::ostringstream oss;
      oss<<"The vector x must be the same length as the number of tips ("<<
        this->ref_tree_.num_tips()<<"), but were"<<input_data.x_.size()<<".";
      throw std::invalid_argument(oss.str());
    } else {
      
      uvec ordNodes = this->ref_tree_.OrderNodes(input_data.names_);
      this->x = At(input_data.x_, ordNodes);
      this->L0 = vec(this->ref_tree_.num_nodes());
      this->L1 = vec(this->ref_tree_.num_nodes());
      this->Q = arma::mat22();
    }
  };

  StateType StateAtRoot() const {
    vec res(2);
    res[0] = L0[this->ref_tree_.num_nodes() - 1];
    res[1] = L1[this->ref_tree_.num_nodes() - 1];
    return res;
  };
  
  void SetParameter(ParameterType const& par) {
    if(par.size() != 2) {
      throw std::invalid_argument(
          "The par vector should be of length 2 with \
      elements corresponding to q01 and q10.");
    }
    if(par[0] <= 0 || par[1] <= 0) {
      throw std::logic_error("The parameters q01 and q10 should be positive.");
    }
    this->q01 = par[0];
    this->q10 = par[1];
    
    this->Q(0,0) = -this->q01;
    this->Q(0,1) = this->q01;
    this->Q(1,0) = this->q10;
    this->Q(1,1) = -this->q10;
    
    using namespace arma;
    cx_vec2 cx_lambdaQ;
    cx_lambdaQ.fill(std::complex<double>(0.0, 0.0));
    cx_mat22 cx_vecQ;
    
    eig_gen(cx_lambdaQ, cx_vecQ, this->Q);
    
    this->lambdaQ = real(cx_lambdaQ);
    this->vecQ = real(cx_vecQ);
    this->vecQ_1 = inv(this->vecQ);
    
  }
  
  inline void InitNode(uint i) {
    if(i < this->ref_tree_.num_tips()) {
      L0[i] = x[i] == 0 ? 1.0 : 0.0;  
      L1[i] = x[i] == 1 ? 1.0 : 0.0;
    } else {
      L0[i] = L1[i] = 1.0;
    }
  }
  
  inline void VisitNode(uint i) {
    
  }
  
  inline void PruneNode(uint i, uint j) {
    double t = this->ref_tree_.LengthOfBranch(i);
    
    using namespace arma;
    // transition probability matrix
    mat22 Pt = this->vecQ * diagmat(exp(this->lambdaQ * t)) * this->vecQ_1;
    
    L0[j] = L0[j] * (Pt(0, 0)*L0[i] + Pt(0, 1)*L1[i]);
    L1[j] = L1[j] * (Pt(1, 0)*L0[i] + Pt(1, 1)*L1[i]);
  }
};
}

#endif //BinaryPoissonModel_H_


src/RCPP__BinaryPoissonModel.cpp
/**
  *  RCPP__BinaryPoissonModel.cpp
  *  SPLITT
  *
  * Copyright 2017 Venelin Mitov
  *
  * This file is part of SPLITT: a generic C++ library for Serial and Parallel
  * Lineage Traversal of Trees.
  *
  * SPLITT is free software: you can redistribute it and/or modify
  * it under the terms of the GNU Lesser General Public License as
  * published by the Free Software Foundation, either version 3 of
  * the License, or (at your option) any later version.
  *
  * SPLITT is distributed in the hope that it will be useful,
  * but WITHOUT ANY WARRANTY; without even the implied warranty of
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  * GNU Lesser General Public License for more details.
  *
  * You should have received a copy of the GNU Lesser General Public
  * License along with SPLITT.  If not, see
  * <http://www.gnu.org/licenses/>.
  *
  * @author Venelin Mitov
  */

#include <Rcpp.h>
#include "./BinaryPoissonModel.h"
    
// [[Rcpp::plugins("cpp11")]]
// [[Rcpp::plugins(openmp)]]
// [[Rcpp::depends("RcppArmadillo")]]

using namespace SPLITT;
using namespace BinaryPoissonUsingSPLITT;

typedef TraversalTask< BinaryPoissonModel<OrderedTree<uint, double>> > TraversalTaskBinaryPoissonModel;


TraversalTaskBinaryPoissonModel* CreateTraversalTaskBinaryPoissonModel(
    Rcpp::List const& tree, uvec const& values) {
  
  Rcpp::IntegerMatrix branches = tree["edge"];
  uvec parents(branches.column(0).begin(), branches.column(0).end());
  uvec daughters(branches.column(1).begin(), branches.column(1).end());
  vec t = Rcpp::as<vec>(tree["edge.length"]);
  uint num_tips = Rcpp::as<Rcpp::CharacterVector>(tree["tip.label"]).size();
  uvec tip_names = Seq(uint(1), num_tips);
  
  typename TraversalTaskBinaryPoissonModel::DataType data(tip_names, values);
  
  return new TraversalTaskBinaryPoissonModel(parents, daughters, t, data);
}


// This will enable returning a copy of the `TraversalAlgorithm`-object stored in
// a `TraversalTaskBinaryPoissonModel` object to a R. This will be used in the MiniBenchmark
// R-function to check things like the OpenMP version used during compilation and
// the number of OpenMP threads at runtime. 
RCPP_EXPOSED_CLASS_NODECL(TraversalTaskBinaryPoissonModel::AlgorithmType)
  
RCPP_MODULE(BinaryPoissonUsingSPLITT__TraversalTaskBinaryPoissonModel) {
  
  // Expose the properties VersionOPENMP and NumOmpThreads from the base 
  // TraversalAlgorithm class
  Rcpp::class_<TraversalTaskBinaryPoissonModel::AlgorithmType::ParentType> (
      "BinaryPoissonUsingSPLITT__BinaryPoissonModel__TraversalAlgorithm"
    )
  .property( "VersionOPENMP",
             &TraversalTaskBinaryPoissonModel::AlgorithmType::ParentType::VersionOPENMP )
  .property( "NumOmpThreads",
             &TraversalTaskBinaryPoissonModel::AlgorithmType::ParentType::NumOmpThreads )
  ;

  // Expose the TraversalTaskBinaryPoissonModel::AlgorithmType specifying that it derives 
  // from the base TraversalAlgorithm class
  Rcpp::class_<TraversalTaskBinaryPoissonModel::AlgorithmType> (
      "BinaryPoissonUsingSPLITT__BinaryPoissonModel__AlgorithmType"
    )
  .derives<TraversalTaskBinaryPoissonModel::AlgorithmType::ParentType>(
      "BinaryPoissonUsingSPLITT__BinaryPoissonModel__TraversalAlgorithm"
    )
  ;
  
  // Finally, expose the TraversalTaskBinaryPoissonModel class - this is the main class in 
  // the module, which will be instantiated from R using the factory function
  // we've just written.
  Rcpp::class_<TraversalTaskBinaryPoissonModel>( "BinaryPoissonUsingSPLITT__TraversalTaskBinaryPoissonModel" )
  // The <argument-type-list> MUST MATCH the arguments of the factory function 
  // defined above.
  .factory<Rcpp::List const&, uvec const&>( &CreateTraversalTaskBinaryPoissonModel )
  // Expose the method that we will use to execute the TraversalTask
  .method( "TraverseTree", &TraversalTaskBinaryPoissonModel::TraverseTree )
  // Expose the algorithm property
  .property( "algorithm", &TraversalTaskBinaryPoissonModel::algorithm )
  ;
}



R/MiniBenchmark.R
# MiniBenchmark.R
# BinaryPoissonUsingSPLITT
# 
# Copyright 2017 Venelin Mitov
# 
# This file is part of SPLITT: a generic C++ library for Serial and Parallel
# Lineage Traversal of Trees.
# 
# SPLITT is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as
# published by the Free Software Foundation, either version 3 of
# the License, or (at your option) any later version.
# 
# SPLITT is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public
# License along with SPLITT.  If not, see
# <http://www.gnu.org/licenses/>.
# 
# @author Venelin Mitov

#' Perform a quick test of the PMM example likelihood calculation
#' @description This function runs a small benchmark to evaluate the SPLITT-package 
#' installation on a given computer. 
#' @param N number of tips in the test phylogenetic tree, default 10000
#' @param Ntests number of calculations within a call to sys.time (the resulting
#' times are averages Ntests calls). Default: 10.
#' @return a data.frame containing the response times for the PMM log-likelihood
#' calculation.
MiniBenchmark <- function(N = 10000, Ntests = 10) {
  
  # needed to pass the check
  speedupX <- NULL
  time.ms <- NULL
  
  cat("Performing a mini-benchmark of the PMM log-likelihood calculation with 
      a tree of size N=", 
      N, ";\nCalling each likelihood calculation Ntests=", 
      Ntests, " times ...\n")
  
  set.seed(10)
  N <- 100
  
  
  x0 <- 1
  q01 <- 0.02
  q10 <- 0.8

  tree <- rtree(N)
  x <- sample(c(0, 1), size = N, replace = TRUE)

    
  if(R.version[['os']]=='linux-gnu') {
    # this only works on linux
    cpuInfo <- system("cat /proc/cpuinfo | grep 'model name' | uniq", intern = TRUE)
  } else if(Sys.info()["sysname"] == "Darwin") {
    # this only works on mac OS x
    cpuInfo <- system("sysctl -a -n machdep.cpu.brand_string", intern = TRUE)
  } else {
    cpuInfo <- "(I don't know how to figure it out on your OS, sorry.)"
  }
  
  cat("CPU: ", cpuInfo, "\n")
  
  ord <- reorder(tree, order = "postorder", index.only = TRUE)
  
  cppPMMObject <- NewBinaryPoissonModelCppObject(x, tree)
  
  if(cppPMMObject$algorithm$VersionOPENMP == 0) {
    cat("It seems that OpenMP was disabled during C++ compilation. For parallel
        tree traversal OpenMP should be available on your system and supported
        by your C++ compiler. Only serial tree traversal is possible with 
        the current SPLITT installation. Please, read the user guide for further 
        instructions on how to build the SPLITT library.")
  } else {
    cat("OpenMP version: ", cppPMMObject$algorithm$VersionOPENMP, "\n")    
  }
  
  cat("Number of threads:",  cppPMMObject$algorithm$NumOmpThreads, "\n")
  
  cat("Measuring calculation times...\n")
  
  # warm-up for mode AUTO
  for(t in 1:1000) BinaryPoissonModelLogLikCpp(x, tree, x0, q01, q10, cppPMMObject, 0)
  
  tR <- system.time(for(t in seq_len(Ntests/10)) BinaryPoissonModelLogLik(x, tree, x0, q01, q10, ord = ord))[3] / (Ntests/10)
  unname(tR)
  
  measureTimeBinaryPoissonModelCpp <- function(mode) {
    unname(
      system.time(
        for(t in seq_len(Ntests)) 
          BinaryPoissonModelLogLikCpp(x, tree, x0, q01, q10, cppPMMObject, mode)
      )[3] / Ntests*1000
    )
  }
  
  resultsBinaryPoissonModel <- rbind(
    data.frame(model = "PMM", 
               mode = "R (serial)", 
               time.ms = tR*1000),
    data.frame(model = "PMM", 
               mode = "C++ (AUTO)", 
               time.ms = measureTimeBinaryPoissonModelCpp(0)),
    data.frame(model = "PMM", 
               mode = "C++ (SINGLE_THREAD_LOOP_POSTORDER)", 
               time.ms = measureTimeBinaryPoissonModelCpp(10)),
    data.frame(model = "PMM", 
               mode = "C++ (SINGLE_THREAD_LOOP_PRUNES)", 
               time.ms = measureTimeBinaryPoissonModelCpp(11)),
    data.frame(model = "PMM", 
               mode = "C++ (SINGLE_THREAD_LOOP_VISITS)", 
               time.ms = measureTimeBinaryPoissonModelCpp(12)),
    data.frame(model = "PMM", 
               mode = "C++ (MULTI_THREAD_LOOP_PRUNES)", 
               time.ms = measureTimeBinaryPoissonModelCpp(21)),
    data.frame(model = "PMM", 
               mode = "C++ (MULTI_THREAD_LOOP_VISITS)", 
               time.ms = measureTimeBinaryPoissonModelCpp(22)),
    data.frame(model = "PMM", 
               mode = "C++ (MULTI_THREAD_LOOP_VISITS_THEN_LOOP_PRUNES)", 
               time.ms = measureTimeBinaryPoissonModelCpp(23)),
    data.frame(model = "PMM", 
               mode = "C++ (MULTI_THREAD_VISIT_QUEUE)", 
               time.ms = measureTimeBinaryPoissonModelCpp(24)),
    data.frame(model = "PMM", 
               mode = "C++ (MULTI_THREAD_LOOP_PRUNES_NO_EXCEPTION)", 
               time.ms = measureTimeBinaryPoissonModelCpp(25)),
    data.frame(model = "PMM", 
               mode = "C++ (HYBRID_LOOP_PRUNES)", 
               time.ms = measureTimeBinaryPoissonModelCpp(31)),
    data.frame(model = "PMM", 
               mode = "C++ (HYBRID_LOOP_VISITS)", 
               time.ms = measureTimeBinaryPoissonModelCpp(32)),
    data.frame(model = "PMM", 
               mode = "C++ (HYBRID_LOOP_VISITS_THEN_LOOP_PRUNES)", 
               time.ms = measureTimeBinaryPoissonModelCpp(33))
  )
  
  rownames(resultsBinaryPoissonModel) <- NULL
  
  resultsBinaryPoissonModel
}
